File: pkg/dart2cpp/sample/dart/hello_test_additional_features.dart
/// Dart 附加特性测试示例
///
/// 包含适配转换器的更多Dart语法特性
///

void main() {
  print('🔥 Dart 附加特性测试开始');

  // 1. 基本数据类型和操作
  testDataTypes();

  // 2. 控制流附加特性
  testAdditionalControlFlow();

  // 3. 函数附加特性
  testAdditionalFunctions();

  // 4. 集合附加特性
  testAdditionalCollections();

  // 5. 类附加特性
  testAdditionalClasses();

  // 6. 泛型附加特性
  testAdditionalGenerics();

  // 7. 异常处理
  testSimpleExceptionHandling();

  // 8. 空安全特性
  testSimpleNullSafety();

  print('✅ Dart 附加特性测试完成');
}

/// 测试数据类型和操作
void testDataTypes() {
  print('\n--- 数据类型和操作 ---');

  // 基本类型
  int age = 25;
  double price = 99.99;
  String name = '张三';
  bool isStudent = true;

  // 字符串插值
  String info = '姓名: $name, 年龄: $age';

  // 字符串方法
  String upper = name.toUpperCase();
  String lower = name.toLowerCase();
  int strLength = name.length;

  // 数值方法
  String ageStr = age.toString();
  int priceInt = price.toInt();

  print('数据类型操作完成');
}

/// 测试附加控制流特性
void testAdditionalControlFlow() {
  print('\n--- 附加控制流特性 ---');

  // switch 语句（更多case）
  int score = 85;
  switch (score) {
    case 100:
    case 99:
    case 98:
      print('满分');
      break;
    case 90:
    case 89:
    case 88:
    case 87:
    case 86:
    case 85:
      print('优秀');
      break;
    case 80:
    case 79:
    case 78:
      print('良好');
      break;
    default:
      print('其他分数');
  }

  // for 循环（不同形式）
  print('倒序循环:');
  for (int i = 5; i >= 1; i--) {
    print('数字: $i');
  }

  // for-in 循环
  List<String> fruits = ['苹果', '香蕉', '橙子'];
  print('水果列表:');
  for (String fruit in fruits) {
    print('  $fruit');
  }

  // while 和 do-while
  int counter = 0;
  while (counter < 3) {
    print('While循环: $counter');
    counter++;
  }

  int doCounter = 0;
  do {
    print('Do-while循环: $doCounter');
    doCounter++;
  } while (doCounter < 3);

  // 条件表达式
  int testAge = 20;
  String description = testAge >= 18 ? '成年人' : '未成年人';
  print('年龄描述: $description');
}

/// 测试附加函数特性
void testAdditionalFunctions() {
  print('\n--- 附加函数特性 ---');

  // 带参数的函数
  String greet(String name) {
    return '你好, $name!';
  }

  // 多参数函数
  int addThreeNumbers(int a, int b, int c) {
    return a + b + c;
  }

  // 调用函数
  String greeting = greet('李四');
  int sum = addThreeNumbers(1, 2, 3);

  print('问候语: $greeting');
  print('三数之和: $sum');
}

/// 测试附加集合特性
void testAdditionalCollections() {
  print('\n--- 附加集合特性 ---');

  // List 操作
  List<int> numbers = [1, 2, 3, 4, 5];
  print('列表长度: ${numbers.length}');
  print('第一个元素: ${numbers.first}');
  print('最后一个元素: ${numbers.last}');

  // List 方法
  numbers.add(6);
  numbers.addAll([7, 8]);
  print('添加元素后长度: ${numbers.length}');

  // Map 操作
  Map<String, int> scores = {'张三': 95, '李四': 87, '王五': 92};
  print('映射大小: ${scores.length}');
  print('张三的分数: ${scores['张三']}');

  // Map 更新
  scores['赵六'] = 88;
  print('更新后映射大小: ${scores.length}');
}

/// 测试附加类特性
void testAdditionalClasses() {
  print('\n--- 附加类特性 ---');

  // 创建对象
  AdditionalPerson person = AdditionalPerson('张三', 25);
  print('创建AdditionalPerson对象: $person');

  // 调用方法
  person.introduce();

  // 访问属性
  print('姓名: ${person.getName()}');
  person.setName('李四');
  print('更新后姓名: ${person.getName()}');

  // 继承
  AdditionalStudent student = AdditionalStudent('王五', 20, '计算机科学');
  print('创建AdditionalStudent对象: $student');
  student.study();
  student.introduce();
}

/// 测试附加泛型特性
void testAdditionalGenerics() {
  print('\n--- 附加泛型特性 ---');

  // 泛型类实例化
  GenericHolder<String> stringHolder = GenericHolder<String>('Hello');
  GenericHolder<int> intHolder = GenericHolder<int>(42);

  print('字符串持有者值: ${stringHolder.getValue()}');
  print('整数持有者值: ${intHolder.getValue()}');

  // 泛型函数调用
  List<int> numbers = [1, 2, 3];
  List<String> strings = ['a', 'b', 'c'];

  int firstNumber = getFirst(numbers);
  String firstString = getFirst(strings);

  print('第一个数字: $firstNumber');
  print('第一个字符串: $firstString');
}

/// 测试简单异常处理
void testSimpleExceptionHandling() {
  print('\n--- 简单异常处理 ---');

  try {
    int result = 10 ~/ 3; // 整数除法
    print('整数除法结果: $result');
  } catch (e) {
    print('捕获到异常');
  }

  print('异常处理完成');
}

/// 测试简单空安全特性
void testSimpleNullSafety() {
  print('\n--- 简单空安全特性 ---');

  // 可空类型
  String? nullableString = 'Hello';
  String? nullString = null;

  // 空值合并操作符
  String nonNull1 = nullableString ?? 'default';
  String nonNull2 = nullString ?? 'default';

  print('空值合并结果: $nonNull1, $nonNull2');

  // 空值安全访问
  List<String>? nullableList = ['a', 'b', 'c'];
  int? length1 = nullableList?.length;
  print('列表长度: $length1');
}

// ========================
// 类定义
// ========================

/// 附加人物类
class AdditionalPerson {
  String _name;
  int _age;

  AdditionalPerson(this._name, this._age);

  // Getter 和 Setter
  String getName() => _name;
  void setName(String value) => _name = value;

  int getAge() => _age;
  void setAge(int value) => _age = value;

  /// 自我介绍
  void introduce() {
    print('大家好，我是 $_name，今年 $_age 岁。');
  }

  @override
  String toString() {
    return 'AdditionalPerson{name: $_name, age: $_age}';
  }
}

/// 附加学生类（继承）
class AdditionalStudent extends AdditionalPerson {
  String _major;

  AdditionalStudent(String name, int age, this._major) : super(name, age);

  // Getter 和 Setter
  String getMajor() => _major;
  void setMajor(String value) => _major = value;

  /// 学习方法
  void study() {
    print('${getName()} 正在学习 $_major');
  }

  @override
  void introduce() {
    print('大家好，我是 ${getName()}，今年 ${getAge()} 岁，专业是 $_major。');
  }

  @override
  String toString() {
    return 'AdditionalStudent{name: ${getName()}, age: ${getAge()}, major: $_major}';
  }
}

/// 泛型持有者类
class GenericHolder<T> {
  T _value;

  GenericHolder(this._value);

  T getValue() => _value;
  void setValue(T newValue) => _value = newValue;
}

// ========================
// 泛型函数
// ========================

/// 获取列表第一个元素的泛型函数
T getFirst<T>(List<T> list) {
  if (list.isEmpty) {
    throw Exception('列表为空');
  }
  return list[0];
}

File: pkg/dart2cpp/sample/cpp_generated/hello_test_additional_features.cpp
#include "dart2cpp.h"

// 工具宏定义

// ============================================================================
// 类: AdditionalPerson
// ============================================================================

class AdditionalPerson {
private:
  String _name;
  Int _age;
public:
  AdditionalPerson(String _name, Int _age) : _name(_name), _age(_age) {
  }
  
  String getName() {
    return this->_name;
  }
  
  String setName(String value) {
    return this->_name = value;
  }
  
  Int getAge() {
    return this->_age;
  }
  
  Int setAge(Int value) {
    return this->_age = value;
  }
  
  Nullable introduce() {
    dart_print(dart_concat(dart_string("大家好，我是 "), (this->_name).toString(), dart_string("，今年 "), (this->_age).toString(), dart_string(" 岁。")));
return Void;
  }
  
  String toString() {
    return dart_concat(dart_string("AdditionalPerson{name: "), (this->_name).toString(), dart_string(", age: "), (this->_age).toString(), dart_string("}"));
  }
  
};

// ============================================================================
// 类: AdditionalStudent
// ============================================================================

class AdditionalStudent : public AdditionalPerson {
private:
  String _major;
public:
  AdditionalStudent(String name, Int age, String _major) : _major(_major), AdditionalPerson(name, age) {
  }
  
  String getMajor() {
    return this->_major;
  }
  
  String setMajor(String value) {
    return this->_major = value;
  }
  
  Nullable study() {
    dart_print(dart_concat((this->getName()).toString(), dart_string(" 正在学习 "), (this->_major).toString()));
return Void;
  }
  
  Nullable introduce() {
    dart_print(dart_concat(dart_string("大家好，我是 "), (this->getName()).toString(), dart_string("，今年 "), (this->getAge()).toString(), dart_string(" 岁，专业是 "), (this->_major).toString(), dart_string("。")));
return Void;
  }
  
  String toString() {
    return dart_concat(dart_string("AdditionalStudent{name: "), (this->getName()).toString(), dart_string(", age: "), (this->getAge()).toString(), dart_string(", major: "), (this->_major).toString(), dart_string("}"));
  }
  
};

// ============================================================================
// 类: GenericHolder
// ============================================================================

template<typename T>
class GenericHolder {
private:
  T _value;
public:
  GenericHolder(T _value) : _value(_value) {
  }
  
  T getValue() {
    return this->_value;
  }
  
  T setValue(T newValue) {
    return this->_value = newValue;
  }
  
};

Nullable testDataTypes();
Nullable testAdditionalControlFlow();
Nullable testAdditionalFunctions();
Nullable testAdditionalCollections();
Nullable testAdditionalClasses();
Nullable testAdditionalGenerics();
Nullable testSimpleExceptionHandling();
Nullable testSimpleNullSafety();
template<typename T>
T getFirst(ObjectPtr<List<T>> list);
Nullable testDataTypes() {
  dart_print(dart_string("\n--- 数据类型和操作 ---"));
auto age = dart_int(25);
auto price = dart_double(99.99);
auto name = dart_string("张三");
auto isStudent = dart_bool(true);
auto info = dart_concat(dart_string("姓名: "), (name).toString(), dart_string(", 年龄: "), (age).toString());
auto upper = name->toUpperCase();
auto lower = name->toLowerCase();
auto strLength = name->size();
auto ageStr = age->toString();
auto priceInt = price->toInt();
dart_print(dart_string("数据类型操作完成"));
return Void;
}

Nullable testAdditionalControlFlow() {
  dart_print(dart_string("\n--- 附加控制流特性 ---"));
auto score = dart_int(85);
if (score == dart_int(100) || score == dart_int(99) || score == dart_int(98)) {
  dart_print(dart_string("满分"));
} else if (score == dart_int(90) || score == dart_int(89) || score == dart_int(88) || score == dart_int(87) || score == dart_int(86) || score == dart_int(85)) {
  dart_print(dart_string("优秀"));
} else if (score == dart_int(80) || score == dart_int(79) || score == dart_int(78)) {
  dart_print(dart_string("良好"));
}else {
  dart_print(dart_string("其他分数"));
}
dart_print(dart_string("倒序循环:"));
for (auto i = dart_int(5); i->operator_greater_equals(dart_int(1)); i = i->operator_sub(dart_int(1))) {
dart_print(dart_string("数字: ") + (i).toString());
}
auto fruits = dart_literal<String>(dart_string("苹果"), dart_string("香蕉"), dart_string("橙子"));
dart_print(dart_string("水果列表:"));
auto sync_for_iterator = fruits->iterator();
for (; sync_for_iterator->hasNext(); ) {
auto fruit = sync_for_iterator->next();
dart_print(dart_string("  ") + (fruit).toString());
}
auto counter = dart_int(0);
while (counter->operator_less(dart_int(3))) {
dart_print(dart_string("While循环: ") + (counter).toString());
counter = counter->operator_add(dart_int(1));
}
auto doCounter = dart_int(0);
do {
dart_print(dart_string("Do-while循环: ") + (doCounter).toString());
doCounter = doCounter->operator_add(dart_int(1));
} while ((doCounter->operator_less(dart_int(3))).toBool());
auto testAge = dart_int(20);
auto description = testAge->operator_greater_equals(dart_int(18)) ? dart_string("成年人") : dart_string("未成年人");
dart_print(dart_string("年龄描述: ") + (description).toString());
return Void;
}

Nullable testAdditionalFunctions() {
  dart_print(dart_string("\n--- 附加函数特性 ---"));
auto greet = [&](String name) { return dart_concat(dart_string("你好, "), (name).toString(), dart_string("!")); };
auto addThreeNumbers = [&](Int a, Int b, Int c) { return a->operator_add(b)->operator_add(c); };
auto greeting = greet(dart_string("李四"));
auto sum = addThreeNumbers(dart_int(1), dart_int(2), dart_int(3));
dart_print(dart_string("问候语: ") + (greeting).toString());
dart_print(dart_string("三数之和: ") + (sum).toString());
return Void;
}

Nullable testAdditionalCollections() {
  dart_print(dart_string("\n--- 附加集合特性 ---"));
auto numbers = dart_literal<Int>(dart_int(1), dart_int(2), dart_int(3), dart_int(4), dart_int(5));
dart_print(dart_string("列表长度: ") + (numbers->size()).toString());
dart_print(dart_string("第一个元素: ") + (numbers->first()).toString());
dart_print(dart_string("最后一个元素: ") + (numbers->last()).toString());
numbers->add(dart_int(6));
numbers->addAll(dart_literal<Int>(dart_int(7), dart_int(8)));
dart_print(dart_string("添加元素后长度: ") + (numbers->size()).toString());
auto scores = Map<String, Int>::createFromEntries({{dart_string("张三"), dart_int(95)}, {dart_string("李四"), dart_int(87)}, {dart_string("王五"), dart_int(92)}});
dart_print(dart_string("映射大小: ") + (scores->size()).toString());
dart_print(dart_string("张三的分数: ") + (scores->operator_index(dart_string("张三"))).toString());
scores->operator_index_set(dart_string("赵六"), dart_int(88));
dart_print(dart_string("更新后映射大小: ") + (scores->size()).toString());
return Void;
}

Nullable testAdditionalClasses() {
  dart_print(dart_string("\n--- 附加类特性 ---"));
auto person = ObjectPtr<AdditionalPerson>(new AdditionalPerson(dart_string("张三"), dart_int(25)));
dart_print(dart_string("创建AdditionalPerson对象: ") + (person).toString());
person->introduce();
dart_print(dart_string("姓名: ") + (person->getName()).toString());
person->setName(dart_string("李四"));
dart_print(dart_string("更新后姓名: ") + (person->getName()).toString());
auto student = ObjectPtr<AdditionalStudent>(new AdditionalStudent(dart_string("王五"), dart_int(20), dart_string("计算机科学")));
dart_print(dart_string("创建AdditionalStudent对象: ") + (student).toString());
student->study();
student->introduce();
return Void;
}

Nullable testAdditionalGenerics() {
  dart_print(dart_string("\n--- 附加泛型特性 ---"));
auto stringHolder = ObjectPtr<GenericHolder<String>>(new GenericHolder<String>(dart_string("Hello")));
auto intHolder = ObjectPtr<GenericHolder<Int>>(new GenericHolder<Int>(dart_int(42)));
dart_print(dart_string("字符串持有者值: ") + (stringHolder->getValue()).toString());
dart_print(dart_string("整数持有者值: ") + (intHolder->getValue()).toString());
auto numbers = dart_literal<Int>(dart_int(1), dart_int(2), dart_int(3));
auto strings = dart_literal<String>(dart_string("a"), dart_string("b"), dart_string("c"));
auto firstNumber = getFirst(numbers);
auto firstString = getFirst(strings);
dart_print(dart_string("第一个数字: ") + (firstNumber).toString());
dart_print(dart_string("第一个字符串: ") + (firstString).toString());
return Void;
}

Nullable testSimpleExceptionHandling() {
  dart_print(dart_string("\n--- 简单异常处理 ---"));
try {
auto result = dart_int(10)->truncatingDivision(dart_int(3));
dart_print(dart_string("整数除法结果: ") + (result).toString());
} catch (const std::exception& e) { /* catch block */ }
// Finally block should be implemented using RAII pattern
dart_print(dart_string("异常处理完成"));
return Void;
}

Nullable testSimpleNullSafety() {
  dart_print(dart_string("\n--- 简单空安全特性 ---"));
auto nullableString = dart_string("Hello");
auto nullString = Null;
auto nonNull1 = dart_null_coalesce(nullableString, dart_string("default"));
auto nonNull2 = dart_null_coalesce(nullString, dart_string("default"));
dart_print(dart_concat(dart_string("空值合并结果: "), (nonNull1).toString(), dart_string(", "), (nonNull2).toString()));
auto nullableList = dart_literal<String>(dart_string("a"), dart_string("b"), dart_string("c"));
auto length1 = dart_null_coalesce(nullableList, Null);
dart_print(dart_string("列表长度: ") + (length1).toString());
return Void;
}

template<typename T>
T getFirst(ObjectPtr<List<T>> list) {
  if (list->isEmpty()) {
throw DartException(Exception::create(dart_string("列表为空")));
}
return list->operator_index(dart_int(0));
}

// ============================================================================
// 主函数
// ============================================================================

int main() {
  try {
    dart_print(dart_string("🔥 Dart 附加特性测试开始"));
testDataTypes();
testAdditionalControlFlow();
testAdditionalFunctions();
testAdditionalCollections();
testAdditionalClasses();
testAdditionalGenerics();
testSimpleExceptionHandling();
testSimpleNullSafety();
dart_print(dart_string("✅ Dart 附加特性测试完成"));
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }
}
